core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java

```diff
@@ -108,17 +99,71 @@ protected final Object invokeNextAndCommitIf1Pc(TxInvocationContext ctx, Prepare
     * becomes a primary owner a new transaction trying to obtain the "real" lock will have to wait for all backup
     * locks to be released. The backup lock will be released either by a commit/rollback/unlock command or by
     * the originator leaving the cluster (if recovery is disabled).
+    *
+    * @return {@code true} if the key was really locked.
+    */
+   protected final boolean lockOrRegisterBackupLock(TxInvocationContext<?> ctx, Object key, long lockTimeout)
+         throws InterruptedException {
+      final Log log = getLog();
+      final boolean trace = log.isTraceEnabled();
+
+      switch (LockUtil.getLockOwnership(key, cdl)) {
+         case PRIMARY:
+            if (trace) {
+               log.tracef("Acquiring locks on %s.", key);
+            }
+            checkPendingAndLockKey(ctx, key, lockTimeout);
+            return true;
+         case BACKUP:
+            if (trace) {
+               log.tracef("Acquiring backup locks on %s.", key);
+            }
+            ctx.getCacheTransaction().addBackupLockForKey(key);
+         default:
+            return false;
+      }
+   }
+
+   /**
+    * Same as {@link #lockOrRegisterBackupLock(TxInvocationContext, Object, long)}
+    *
+    * @return a collection with the keys locked.
     */
-   protected final void lockAndRegisterBackupLock(TxInvocationContext ctx, Object key, long lockTimeout, boolean skipLocking) throws InterruptedException {
-      //with DeltaCompositeKey, the locks should be acquired in the owner of the delta aware key.
-      Object keyToCheck = key instanceof DeltaCompositeKey ?
-            ((DeltaCompositeKey) key).getDeltaAwareValueKey() :
-            key;
-      if (cdl.localNodeIsPrimaryOwner(keyToCheck)) {
-         lockKeyAndCheckOwnership(ctx, key, lockTimeout, skipLocking);
-      } else if (cdl.localNodeIsOwner(keyToCheck)) {
-         ctx.getCacheTransaction().addBackupLockForKey(key);
+   protected final Collection<Object> lockAllOrRegisterBackupLock(TxInvocationContext<?> ctx, Collection<Object> keys,
+                                                                  long lockTimeout) throws InterruptedException {
+      if (keys.isEmpty()) {
+         return Collections.emptyList();
       }
+
+      final Log log = getLog();
+      final boolean trace = log.isTraceEnabled();
+
+      Collection<Object> keysToLock = new ArrayList<>(keys.size());
+
+      for (Object key : keys) {
```

Comment:
It is not needed. I find out that the visit* methods were added the key to the affected set. So, it is redundant to add it here :)